opal/mca/common/ofi/common_ofi.c

```diff
@@ -482,10 +482,16 @@ static int compute_dev_distances(pmix_device_distance_t **distances,
     if (PMIX_SUCCESS != ret) {
         goto out;
     }
+    /* if we are not bound, then we cannot compute distances */
+    if (hwloc_bitmap_iszero(cpuset.bitmap) ||
```

Comment:
Been thinking about this. I think we should move this check in get_nearest_nics() before we call PMIx_Get() to get the distances. 

```
diff --git a/opal/mca/common/ofi/common_ofi.c b/opal/mca/common/ofi/common_ofi.c
index e882c3c833..acba2b0a91 100644
--- a/opal/mca/common/ofi/common_ofi.c
+++ b/opal/mca/common/ofi/common_ofi.c
@@ -525,8 +525,15 @@ get_nearest_nics(int *num_distances, pmix_value_t **valin)
     uint16_t near = USHRT_MAX;
     pmix_info_t directive;
     pmix_value_t *val = NULL;
+    pmix_cpuset_t cpuset;
     pmix_device_distance_t *distances, *nearest = NULL;
 
+    PMIX_CPUSET_CONSTRUCT(&cpuset);
+    if (hwloc_bitmap_iszero(cpuset.bitmap) ||
+        hwloc_bitmap_isfull(cpuset.bitmap)) {
+        goto out
+    }
+
     PMIx_Info_load(&directive, PMIX_OPTIONAL, NULL, PMIX_BOOL);
     ret = PMIx_Get(&opal_process_info.myprocid,
              PMIX_DEVICE_DISTANCES, &directive, 1, &val);
```
I believe this will avoid us having issues if someone tries to initialize MPI without being bound.

what do you think?

Reply:
I'm afraid that won't work - cpuset_construct doesn't populate the bitmap. You have to retrieve the cpuset from PMIx before you can test for `iszero` or `isfull`. I'm not sure what you are trying to accomplish as the logic in the PR is protecting us against the "not bound" case.

Reply:
yes I missed the actual building of the cpuset. I'm trying to avoid this:

```
#0  hwloc_get_cpubind (topology=0x0, set=0x74f7a0, flags=2) at ../../../../../source/hwloc-2.7.1/hwloc/bind.c:112
#1  0x00007fffe0d1b4b1 in pmix_hwloc_get_cpuset (cpuset=0x7fffffff6df0, ref=1 '\001')
    at ../../../../../../../source/openmpi-main-Apr5_23/3rd-party/openpmix/src/hwloc/pmix_hwloc.c:1002
#2  0x00007fffe0d4c36c in PMIx_Get_cpuset (cpuset=0x7fffffff6df0, ref=1 '\001')
    at ../../../../../../../source/openmpi-main-Apr5_23/3rd-party/openpmix/src/client/pmix_client_topology.c:89
#3  0x00007fffed9683df in compute_dev_distances (distances=0x7fffffff6e40, ndist=0x7fffffff7080)
    at ../../../../../../../../source/openmpi-main-Apr5_23/opal/mca/common/ofi/common_ofi.c:480
#4  0x00007fffed968565 in get_nearest_nics (num_distances=0x7fffffff70dc, valin=0x7fffffff70e8)
    at ../../../../../../../../source/openmpi-main-Apr5_23/opal/mca/common/ofi/common_ofi.c:534
#5  0x00007fffed969104 in opal_common_ofi_select_provider (provider_list=0x86ccb0, process_info=0x7fffed9cda40 <opal_process_info>)
    at ../../../../../../../../source/openmpi-main-Apr5_23/opal/mca/common/ofi/common_ofi.c:800
#6  0x00007fffeb9079cb in select_ofi_provider (providers=0x86cf50, include_list=0x0, exclude_list=0x70ee30)
    at ../../../../../../../../source/openmpi-main-Apr5_23/ompi/mca/mtl/ofi/mtl_ofi_component.c:357
#7  0x00007fffeb908dbc in ompi_mtl_ofi_component_init (enable_progress_threads=false, enable_mpi_threads=false, accelerator_support=0x7fffebad4cb0 <mca_mtl_ofi_component+272>)
    at ../../../../../../../../source/openmpi-main-Apr5_23/ompi/mca/mtl/ofi/mtl_ofi_component.c:782
#8  0x00007fffeb8fe9a7 in ompi_mtl_base_select (enable_progress_threads=false, enable_mpi_threads=false, priority=0x7fffffff7324)
    at ../../../../../../../source/openmpi-main-Apr5_23/ompi/mca/mtl/base/mtl_base_frame.c:78
#9  0x00007fffeb9e8e3d in mca_pml_cm_component_init (priority=0x7fffffff7324, enable_progress_threads=false, enable_mpi_threads=false)
    at ../../../../../../../../source/openmpi-main-Apr5_23/ompi/mca/pml/cm/pml_cm_component.c:146
#10 0x00007fffeb9c571f in mca_pml_base_select (enable_progress_threads=false, enable_mpi_threads=false)
    at ../../../../../../../source/openmpi-main-Apr5_23/ompi/mca/pml/base/pml_base_select.c:127
#11 0x00007fffeb76af14 in ompi_mpi_instance_init_common (argc=2, argv=0x7fffffff81b8) at ../../../../../source/openmpi-main-Apr5_23/ompi/instance/instance.c:508
#12 0x00007fffeb76b7e5 in ompi_mpi_instance_init (ts_level=0, info=0x419b40 <ompi_mpi_info_null>, errhandler=0x7fffebadd520 <ompi_mpi_errors_are_fatal>, 
    instance=0x7fffebaed580 <ompi_mpi_instance_default>, argc=2, argv=0x7fffffff81b8) at ../../../../../source/openmpi-main-Apr5_23/ompi/instance/instance.c:814
#13 0x00007fffeb75df7b in ompi_mpi_init (argc=2, argv=0x7fffffff81b8, requested=0, provided=0x7fffffff7f4c, reinit_ok=false)
    at ../../../../../source/openmpi-main-Apr5_23/ompi/runtime/ompi_mpi_init.c:359
#14 0x00007fffeb7b54ca in PMPI_Init (argc=0x7fffffff7f8c, argv=0x7fffffff7f80) at ../../../../../../../source/openmpi-main-Apr5_23/ompi/mpi/c/init.c:67
#15 0x0000000000402331 in main (argc=2, argv=0x7fffffff81b8) at osu_bibw.c:58

```

Reply:
Actually we'll get this crash irregardless, because it's in the PMIx_Get_cpuset(). I guess the goal is to avoid using hwloc if it's not initialized. 

```
(gdb) l
107           int err = topology->binding_hooks.get_thisproc_cpubind(topology, set, flags);
108           if (err >= 0 || errno != ENOSYS)
109             return err;
110           /* ENOSYS, fallback */
111         }
112         if (topology->binding_hooks.get_thisthread_cpubind)
113           return topology->binding_hooks.get_thisthread_cpubind(topology, set, flags);
114       }
115
116       errno = ENOSYS;
(gdb) p topology
$1 = (hwloc_topology_t) 0x0
```
Is there a way to ensure that the hwloc topology is initialized before performing any of these operations?

Reply:
I assume PMIx_Init has already been called - yes? If so, then it looks like some kind of bug in PMIx as the topology must be defined. If not, then it's a bug in OMPI as you cannot invoke PMIx functions without first calling PMIx_Init.

Reply:
As far as I can tell the place which most likely calls PMIx_Init() in OMPI is here:

```
int ompi_rte_init(int *pargc, char ***pargv)
...
    /* initialize the selected module */
    if (!PMIx_Initialized() && (PMIX_SUCCESS != (ret = PMIx_Init(&opal_process_info.myprocid, NULL, 0)))) {
        /* if we get PMIX_ERR_UNREACH indicating that we cannot reach the
         * server, then we assume we are operating as a singleton */
        if (PMIX_ERR_UNREACH == ret) {
```
gdbing this I see that it is indeed getting called:

```
Breakpoint 1, ompi_rte_init (pargc=0x7fffffff73ac, pargv=0x7fffffff73a0) at ../../../../../source/openmpi-main-Apr5_23/ompi/runtime/ompi_rte.c:531
531         char *error = NULL;
(gdb) b PMIx_Init
Breakpoint 2 at 0x7fffe0d2147b: file ../../../../../../../source/openmpi-main-Apr5_23/3rd-party/openpmix/src/client/pmix_client.c, line 521.
(gdb) c
Continuing.
[New Thread 0x7fffdb6ff700 (LWP 118214)]
[New Thread 0x7ff7c31ff700 (LWP 118215)]
[Thread 0x7ff7c31ff700 (LWP 118215) exited]

Thread 1 "osu_bibw" hit Breakpoint 2, PMIx_Init (proc=0x7fffed9cda48 <opal_process_info+8>, info=0x0, ninfo=0) at ../../../../../../../source/openmpi-main-Apr5_23/3rd-party/openpmix/src/client/pmix_client.c:521
521         pmix_status_t rc = PMIX_SUCCESS;
(gdb) c
Continuing.
[New Thread 0x7ff7a35ff700 (LWP 118218)]

Thread 1 "osu_bibw" hit Breakpoint 2, PMIx_Init (proc=0x0, info=0x7fffffff6ae0, ninfo=4) at ../../../../../../../source/openmpi-main-Apr5_23/3rd-party/openpmix/src/client/pmix_client.c:521
521         pmix_status_t rc = PMIX_SUCCESS;
(gdb) c
Continuing.

Thread 1 "osu_bibw" received signal SIGSEGV, Segmentation fault.
hwloc_get_cpubind (topology=0x0, set=0x74f7a0, flags=2) at ../../../../../source/hwloc-2.7.1/hwloc/bind.c:112
112         if (topology->binding_hooks.get_thisthread_cpubind)
```
Haven't delved into further, but it looks like maybe there should be some protection in PMIx in order not to pass a NULL value in the below code

```
#1  0x00007fffe0d1b4b1 in pmix_hwloc_get_cpuset (cpuset=0x7fffffff6df0, ref=1 '\001')
    at ../../../../../../../source/openmpi-main-Apr5_23/3rd-party/openpmix/src/hwloc/pmix_hwloc.c:1002
1002            rc = hwloc_get_cpubind(pmix_globals.topology.topology, cpuset->bitmap, flag);
(gdb) l
997
998         cpuset->bitmap = hwloc_bitmap_alloc();
999         if (NULL != testcpuset) {
1000            rc = hwloc_bitmap_sscanf(cpuset->bitmap, testcpuset);
1001        } else {
1002            rc = hwloc_get_cpubind(pmix_globals.topology.topology, cpuset->bitmap, flag);
1003        }
1004        if (0 != rc) {
1005            hwloc_bitmap_free(cpuset->bitmap);
1006            return PMIX_ERR_TAKE_NEXT_OPTION;

```

Reply:
@amirshehataornl You had requested changes here a few weeks ago, but I think much has changed since then.  Do you have any current objections to merging now?